### PR TITLE
246 sha tables

### PR DESCRIPTION
**Pull request by Iluvmagick**

### CODE REVIEW

File: include/nil/blueprint/components/hashes/sha2/plonk/sha256_process.hpp [hunk 1]
```diff
@@ -407,71 +474,91 @@ namespace nil {
...
+                        var(component.W(3), 0), var(component.W(4), 0)};
+                    const std::array<var, 4> b_chunks_1 = {
+                        var(component.W(1), -1), var(component.W(2), -1),
+                        var(component.W(3), -1), var(component.W(4), -1)};
+                    const std::array<std::array<var, 4>, 2> b_chunks = {b_chunks_0, b_chunks_1};
+
+                    const std::array<var, 4> sigma1_chunks_1 = {
+                        var(component.W(5), -1), var(component.W(6), -1),
+                        var(component.W(7), -1), var(component.W(8), -1)};
+                    const std::array<var, 4> a2_chunks = {
```
Comment (Luannet) [inline include/nil/blueprint/components/hashes/sha2/plonk/sha256_process.hpp:497]: var(component.W(8), 0) is integral_a2 , not a chunk

File: include/nil/blueprint/components/hashes/sha2/plonk/sha256_process.hpp [hunk 2]
```diff
@@ -486,120 +573,181 @@ namespace nil {
...
+                    lookup_constraint_type lookup_constraint_10 =
+                        {lookup_tables_indices.at("sha256_reverse_sparse_base4/full"),
+                         {Sigma0_chunks[0][2], Sigma0_chunks[1][2]}};
+                    lookup_constraint_type lookup_constraint_11 =
+                        {lookup_tables_indices.at("sha256_reverse_sparse_base4/full"),
+                         {Sigma0_chunks[0][3], Sigma0_chunks[1][3]}};
+
+                    lookup_constraint_type lookup_constraint_12 =
+                        {lookup_tables_indices.at("sha256_sparse_base4/first_column"),
+                         {var(component.W(6), 1)}};
```
Comment (Luannet) [inline include/nil/blueprint/components/hashes/sha2/plonk/sha256_process.hpp:735]: change to a_new_chunks

File: include/nil/blueprint/components/hashes/sha2/plonk/sha256_process.hpp [hunk 3]
```diff
@@ -628,77 +776,98 @@ namespace nil {
...
-                                           var(component.W(4), -1) * (1 << 11) + var(component.W(5), -1) * (1 << 25));
+                            (e_chunks[0][0] + e_chunks[0][1] * (1 << 6) +
+                             e_chunks[0][2] * (1 << 11) + e_chunks[0][3] * (1 << 25));
                     auto constraint_2 =
                         var(component.W(0), 0) -
-                        (var(component.W(1), -1) + var(component.W(2), 0) * base7_value.pow(6) +
-                         var(component.W(3), 0) * base7_value.pow(11) + var(component.W(4), 0) * base7_value.pow(25));
+                            (e_chunks[1][0] + e_chunks[1][1] * base7_value.pow(6) +
+                             e_chunks[1][2] * base7_value.pow(11) + e_chunks[1][3] * base7_value.pow(25));
                     // auto constraint_3 =
```
Comment (Luannet) [inline include/nil/blueprint/components/hashes/sha2/plonk/sha256_process.hpp:812]: remove comment

File: include/nil/blueprint/components/hashes/sha2/plonk/sha256_process.hpp [hunk 4]
```diff
@@ -628,77 +776,98 @@ namespace nil {
...
+                    lookup_constraint lookup_constraint_10 =
+                        {lookup_tables_indices.at("sha256_reverse_sparse_base7/full"),
+                         {Sigma1_chunks[0][2], Sigma1_chunks[1][2]}};
+                    lookup_constraint lookup_constraint_11 =
+                        {lookup_tables_indices.at("sha256_reverse_sparse_base7/full"),
+                         {Sigma1_chunks[0][3], Sigma1_chunks[1][3]}};
 
                     selectors[1] = bp.add_lookup_gate({
-//                        lookup_constraint_1,
+                        /*lookup_constraint_1,
```
Comment (Luannet) [inline include/nil/blueprint/components/hashes/sha2/plonk/sha256_process.hpp:864]: uncomment

**Review by Luannet — CHANGES_REQUESTED** [review]

File: include/nil/blueprint/components/hashes/sha2/plonk/sha256_process.hpp [hunk 5]
```diff
@@ -407,71 +474,91 @@ namespace nil {
...
                     circuit<crypto3::zk::snark::plonk_constraint_system<BlueprintFieldType, ArithmetizationParams>> &bp,
                     assignment<crypto3::zk::snark::plonk_constraint_system<BlueprintFieldType, ArithmetizationParams>> &assignment,
-                    const std::map<std::string, std::size_t> &lookup_tables_indices
+                    const typename lookup_library<BlueprintFieldType>::left_reserved_type &lookup_tables_indices
                 ) {
                     std::array<std::size_t, 2> selectors;
                     using var = typename plonk_sha256_process<BlueprintFieldType, ArithmetizationParams>::var;
+                    using lookup_constraint = crypto3::zk::snark::plonk_lookup_constraint<BlueprintFieldType>;
+
+                    const std::array<var, 4> a_chunks_0_next = {
```
[the remark this answers is not in the document]
Comment (Iluvmagick) [reply include/nil/blueprint/components/hashes/sha2/plonk/sha256_process.hpp]: Removed.

File: include/nil/blueprint/components/hashes/sha2/plonk/sha256_process.hpp [hunk 6]
```diff
@@ -181,17 +181,39 @@ namespace nil {
...
                             this->_table[1].push_back(value_chunks[1][0]);
                         }
-                        std::cout << "=============================" << std::endl;
                     }
 
                     virtual std::size_t get_columns_number(){return 2;}
                     virtual std::size_t get_rows_number(){return 16384;}
                 };
 
+                class reverse_sparse_values_base4_table : public lookup_table_definition {
```
[the remark this answers is not in the document]
Comment (Iluvmagick) [reply include/nil/blueprint/components/hashes/sha2/plonk/sha256_process.hpp]: Done.

File: include/nil/blueprint/components/hashes/sha2/plonk/sha256_process.hpp (hunk 1 above)
Comment (Iluvmagick) [reply include/nil/blueprint/components/hashes/sha2/plonk/sha256_process.hpp]: Renamed here and in another place.

File: include/nil/blueprint/components/hashes/sha2/plonk/sha256_process.hpp (hunk 2 above)
Comment (Iluvmagick) [reply include/nil/blueprint/components/hashes/sha2/plonk/sha256_process.hpp]: Changed.

File: include/nil/blueprint/components/hashes/sha2/plonk/sha256_process.hpp (hunk 4 above)
Comment (Iluvmagick) [reply include/nil/blueprint/components/hashes/sha2/plonk/sha256_process.hpp]: Uncommented, and corrected constraints.

File: include/nil/blueprint/components/hashes/sha2/plonk/sha256_process.hpp (hunk 3 above)
Comment (Iluvmagick) [reply include/nil/blueprint/components/hashes/sha2/plonk/sha256_process.hpp]: Removed.

**Pull request merged**